File: frontend/lib/models/job.dart
import 'package:frontend/models/User.dart';
import 'package:frontend/models/contact.dart';
import 'package:frontend/models/tasks.dart';

class Job {
  int? id;
  String? title;
  String? description;
  String? status;
  String? priority;
  DateTime? scheduledDate;
  DateTime? completionDate;
  double? estimatedCost;
  double? actualCost;
  String? notes;
  DateTime? dateCreated;
  DateTime? dateUpdated;
  int? contactId;
  Contact? contact;
  String? assignedUserId;
  User? assignedUser;
  String? createdByUserId;
  User? createdByUser;
  List<Tasks>? tasks;

  Job({
    this.id,
    this.title,
    this.description,
    this.status,
    this.priority,
    this.scheduledDate,
    this.completionDate,
    this.estimatedCost,
    this.actualCost,
    this.notes,
    this.dateCreated,
    this.dateUpdated,
    this.contactId,
    this.contact,
    this.assignedUserId,
    this.assignedUser,
    this.createdByUserId,
    this.createdByUser,
    this.tasks,
  });

  factory Job.fromJson(Map<String, dynamic> json) {
    return Job(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      status: json['status'],
      priority: json['priority'],
      scheduledDate: DateTime.parse(json['scheduledDate']),
      completionDate: json['completionDate'] != null
          ? DateTime.parse(json['completionDate'])
          : null,
      estimatedCost: json['estimatedCost'] != null
          ? json['estimatedCost'].toDouble()
          : null,
      actualCost: json['actualCost'] != null
          ? json['actualCost'].toDouble()
          : null,
      notes: json['notes'],
      dateCreated: DateTime.parse(json['dateCreated']),
      dateUpdated: json['dateUpdated'] != null
          ? DateTime.parse(json['dateUpdated'])
          : null,
      contactId: json['contactId'],
      contact: json['contact'] != null ? Contact.fromJson(json['contact']) : null,
      assignedUserId: json['assignedUserId'],
      assignedUser: json['assignedUser'] != null
          ? User.fromJson(json['assignedUser'])
          : null,
      createdByUserId: json['createdByUserId'],
      createdByUser: json['createdByUser'] != null
          ? User.fromJson(json['createdByUser'])
          : null,
      tasks: (json['tasks'] as List?)
          ?.map((item) => Tasks.fromJson(item))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'status': status,
      'priority': priority,
      'scheduledDate': scheduledDate?.toIso8601String(),
      'completionDate': completionDate?.toIso8601String(),
      'estimatedCost': estimatedCost,
      'actualCost': actualCost,
      'notes': notes,
      'dateCreated': dateCreated?.toIso8601String(),
      'dateUpdated': dateUpdated?.toIso8601String(),
      'contactId': contactId,
      'contact': contact?.toJson(),
      'assignedUserId': assignedUserId,
      'assignedUser': assignedUser?.toJson(),
      'createdByUserId': createdByUserId,
      'createdByUser': createdByUser?.toJson(),
      'tasks': tasks?.map((e) => e.toJson()).toList(),
    };
  }
}

File: backend/backend/Areas/Main/Models/Job.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using backend.Areas.Identity.Models;
using backend.Areas.Main.Models.Enums;

namespace backend.Areas.Main.Models;

public class Job
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }

    [Required]
    [StringLength(150)]
    public string Title { get; set; }

    [StringLength(500)]
    public string? Description { get; set; }

    [Required]
    [StringLength(100)]
    public Status Status { get; set; }

    [Required]
    [StringLength(50)]
    public Priority Priority { get; set; }

    public DateTime ScheduledDate { get; set; }

    public DateTime? CompletionDate { get; set; }

    [Column(TypeName = "decimal(18,2)")]
    public decimal EstimatedCost { get; set; }

    [Column(TypeName = "decimal(18,2)")]
    public decimal? ActualCost { get; set; }

    public string? Notes { get; set; }

    public DateTime DateCreated { get; set; } = DateTime.UtcNow;
    public DateTime? DateUpdated { get; set; }
    public int ContactId { get; set; }
    [ForeignKey(nameof(ContactId))]
    public virtual Contact Contact { get; set; }
    public string? AssignedUserId { get; set; }
    [ForeignKey(nameof(AssignedUserId))]
    public virtual User AssignedUser { get; set; }
    public string? CreatedByUserId { get; set; }
    [ForeignKey(nameof(CreatedByUserId))]
    public User CreatedByUser { get; set; }
    public IEnumerable<Tasks>? Tasks { get; set; }
    public IEnumerable<JobNotes>? JobNotes { get; set; }
    public IEnumerable<JobTask>? JobTasks { get; set; }
}